[USMON-1439] - Add error types to USM Redis stats

Saarshta opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
Adds reporting of error prefix for USM Redis stats.
### Motivation
Let customers know what type of errors they receive, other than indicating whether an error occurred or not.
### Describe how you validated your changes
Updated tests to use error types.
Tested in staging.
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Commit:
WIP - report Redis error types

Commit:
Merge branch 'main' into saar.shtalryd/redis-errors

Commit:
WIP

Commit:
Minor bug fix to propagate error type properly

Commit:
Fix tests, merge main

Commit:
dda inv tidy

Commit:
Fix tests

Commit:
Fix wrong redis error type

Commit:
Fix tests

Commit:
Minor bug fix

Commit:
Reorder redis_transaction_t struct

Commit:
Fix tests

Commit:
Fix tests

Commit:
Fix tests

Commit:
Fix tests

Commit:
Fix verifier

Commit:
Fix verifier

Commit:
Reduce memory usage

Commit:
Merge main

Commit:
Bump agent-payload version

guyarb commented on pkg/network/ebpf/c/protocols/redis/decoding.h:
```diff
@@ -209,13 +235,47 @@ static __always_inline void redis_batch_enqueue_wrapper(conn_tuple_t *tuple, red
     if (first_byte == RESP_ERROR_PREFIX) {
         transaction->is_error = true;
+        char error_prefix[MAX_ERROR_SIZE] = {};
+        bool got_error = extract_redis_error_prefix(pkt, error_prefix, sizeof(error_prefix));
+        if (got_error) {
+            transaction->error = map_redis_error_prefix(error_prefix);
+        } else {
+            transaction->error = REDIS_ERR_UNKNOWN;
+        }
```
With my previous comment, I think the condition is redundant, as eventually `transaction->error = map_redis_error_prefix` should suffice, as the default behavior is `REDIS_ERR_UNKNOWN`

guyarb commented on pkg/network/encoding/marshal/usm_redis.go:
```diff
@@ -110,3 +106,60 @@ func (e *redisEncoder) Close() {
 	e.byConnection.Close()
 }
+
+func mapRedisErrorType(err redis.ErrorType) int32 {
```
Since the mapping is static, I'd migrate it into a map

guyarb commented on pkg/network/protocols/redis/model_linux.go:
```diff
@@ -110,3 +122,60 @@ func (c CommandType) String() string {
 	}
 }
+
+func (e ErrorType) String() string {
```
Same comment as above, consider using a map

guyarb reviewed (commented):
reviewed

Commit:
Add debug log for redis errors

Commit:
Add debug log for redis errors

Commit:
attempt to fix cgo alignment

Commit:
attempt to fix cgo alignment

Commit:
Remove debug logs

Commit:
CR comments

guyarb committed:
Merge branch 'main' into saar.shtalryd/redis-errors

guyarb committed:
usm: redis: Remove redundant is_error from the kernel struct

As we now report the actual error, we don't need the binary flag to indicate if we have an error

guyarb committed:
usm: redis: Added debug log for finding the error on staging

guyarb committed:
usm: merge with above: remove is_error from tests

guyarb committed:
Wip
